### ISSUE TITLE
panic: first replica is not the primary replica for the token

### ISSUE THREAD

**Issue by martin-sucha** [gocql/gocql#1274]
### What version of Cassandra are you using?

ScyllaDB 2018.1.10

### What version of Gocql are you using?

github.com/martin-sucha/gocql@v0.0.0-20190227121127-86b0b61a5e6c

### What did you do?

After our database cluster went dead due to a server-side database bug, the gocql client panicked.

### What did you expect to see?

I'd expect the client to be resilient against cluster failures. If it can't reconnect to the nodes, it should not panic, but return errors instead, so that the service can fail gracefully.

### What did you see instead?

The gocql driver panicked with:

```
github.com/gocql/gocql.(*networkTopology).replicaMap(0xc00101e090, 0xc00101e078, 0x0, 0x0, 0xc007394000, 0x100, 0x100, 0x4634f5)
	/go/pkg/mod/github.com/martin-sucha/gocql@v0.0.0-20190227121127-86b0b61a5e6c/topology.go:203 +0x1291
github.com/gocql/gocql.(*tokenAwareHostPolicy).updateKeyspaceMetadata(0xc00017cf80, 0xc0000380f1, 0x7)
	/go/pkg/mod/github.com/martin-sucha/gocql@v0.0.0-20190227121127-86b0b61a5e6c/policies.go:447 +0x395
github.com/gocql/gocql.(*tokenAwareHostPolicy).AddHost(0xc00017cf80, 0xc0003e02c0)
	/go/pkg/mod/github.com/martin-sucha/gocql@v0.0.0-20190227121127-86b0b61a5e6c/policies.go:478 +0x6e
github.com/gocql/gocql.(*Session).addNewNode(0xc0000edc00, 0xc0003e02c0)
	/go/pkg/mod/github.com/martin-sucha/gocql@v0.0.0-20190227121127-86b0b61a5e6c/events.go:188 +0x9f
github.com/gocql/gocql.(*Session).handleNodeUp(0xc0000edc00, 0xc000b6eab0, 0x10, 0x10, 0x2352, 0xc00352ce01)
	/go/pkg/mod/github.com/martin-sucha/gocql@v0.0.0-20190227121127-86b0b61a5e6c/events.go:273 +0x10d
github.com/gocql/gocql.(*Session).reconnectDownedHosts(0xc0000edc00, 0x12a05f200)
	/go/pkg/mod/github.com/martin-sucha/gocql@v0.0.0-20190227121127-86b0b61a5e6c/session.go:284 +0x112
created by github.com/gocql/gocql.(*Session).init
	/go/pkg/mod/github.com/martin-sucha/gocql@v0.0.0-20190227121127-86b0b61a5e6c/session.go:233 +0x4a5
```

**Comment by martin-sucha** [comment]
I think this might be caused by a race condition in accessing `tokenAwareHostPolicy.hosts` and `tokenAwareHostPolicy.tokenRing`. The `networkTopology.replicaMap` assumes consistent hosts and tokenRing, but it might be called with empty hosts and some tokens due to the race condition, hence hitting the panic.

I think the change ec4793573d1447b6f92a1b359a0594566fad9d0e by @mmatczuk might be related in that it might make the race condition more likely to be hit as `updateKeyspaceMetadata` is called more often after this change. It might be even called concurrently from goroutine started in Session.init at `session.go:233` (see trace in the issue description) and in Session.init on line 256 via call to `s.policy.KeyspaceChanged` which calls `updateKeyspaceMetadata`.

One sequence of events that might lead to the race condition & panic could be like below (not sure whether it is related to the actual trace in the issue description though):

A: calls `tokenAwareHostPolicy.AddHost`
A: calls `tokenAwareHostPolicy.HostUp`
A: calls `tokenAwareHostPolicy.updateKeyspaceMetadata`
A: calls `tr, _ := t.tokenRing.Load().(*tokenRing)` to read token Ring
B: calls `tokenAwareHostPolicy.RemoveHost`
B: calls `tokenAwareHostPolicy.HostDown` which calls `t.hosts.remove(host.ConnectAddress())`
A: calls `strat.replicaMap(t.hosts.get(), tr.tokens)`

So, is the correct fix to syncronize acess to `t.hosts` and `t.tokenRing`?

**Comment by mmatczuk** [comment]
Well it's complicated... maybe @Zariel can shed some light on the design of `tokenAwareHostPolicy` and synchronization in particular. We have `mu` rw mutex to protect hosts that is `cowList` that uses atomics internally. We also have `tokenRing` and `keyspaces` that are atomics, I guess that was introduced for performance reasons, otherwise I do not see why `mu` shall not synchronized all the state?

**Comment by javiervazquezh** [comment]
I'm getting the same exception

Cassandra 3.11.4

github.com/gocql/gocql.(*networkTopology).replicaMap(0xc00035c020, 0xc00055fc80, 0x4, 0x4, 0xc000618000, 0x500, 0x555, 0x401845)
	/apps/valrs/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20190423091413-b99afaf3b163/topology.go:203 +0x1366
github.com/gocql/gocql.(*tokenAwareHostPolicy).updateKeyspaceMetadata(0xc000044980, 0xc00002ae90, 0x9)
	/apps/valrs/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20190423091413-b99afaf3b163/policies.go:447 +0x395
github.com/gocql/gocql.(*tokenAwareHostPolicy).AddHost(0xc000044980, 0xc00044a160)
	/apps/valrs/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20190423091413-b99afaf3b163/policies.go:478 +0x6e
github.com/gocql/gocql.(*Session).addNewNode(0xc00012d180, 0xc00044a160)
	/apps/valrs/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20190423091413-b99afaf3b163/events.go:188 +0x9f
github.com/gocql/gocql.(*Session).handleNodeUp(0xc00012d180, 0xc0002ee510, 0x10, 0x10, 0x23b6, 0x1)
	/apps/valrs/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20190423091413-b99afaf3b163/events.go:273 +0x113
github.com/gocql/gocql.(*Session).reconnectDownedHosts(0xc00012d180, 0xdf8475800)
	/apps/valrs/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20190423091413-b99afaf3b163/session.go:284 +0x112
created by github.com/gocql/gocql.(*Session).init
	/apps/valrs/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20190423091413-b99afaf3b163/session.go:233 +0x4a5

Is there a workaround for this issue?

**Comment by martin-sucha** [comment]
I have a reproducer at https://github.com/kiwicom/gocql/commit/8fe7e93f55c9758f4a2a771771edf4a938e3bb63

**Comment by martin-sucha** [comment]
I have work-in-progress patch, will send pull request next week.

**Comment by martin-sucha** [comment]
PR here: https://github.com/gocql/gocql/pull/1332

**Comment by rodgarrison** [comment]
I hit this issue now about every 5 mins - my task basically won't stay up. Although I made minor changes to the queries where I found the bind/iter calls didn't match the select statements, fixing those didn't help. On the other hand I have a POC tasks that connects to cassandra and dumps one table. Even when when multi-threaded, I cannot get it to panic. Unfortunately my production tasks does. My panic traceback is identical to OP's.
